### ISSUE TITLE
Feasibility to have Folder structure in boardfarm to manage testcase files

### ISSUE THREAD

**Issue by Parumugam** [lgirdk/boardfarm#125]
Hello,

As we have huge number of testcases, its not good to write the code in single file as it will grow in space and also, its creating "Merge Conflict" if atleast 2 engineers are working on the same file in parallel. 

Another idea would be to create multiple files (Ex: management_snmp_p1.py, management_snmp_p2.py, etc..), but still we have lot many P2 and P3 cases and hence these files will also grow in space and chances of Merge Conflict is still there.

So, shall we please think of having folders, so that folks can create single file of code per testcase which shall be saved in a particular folder path (Example: Management--> snmp --> P1 --> Testcase_1.py; Management--> snmp --> P1 --> Testcase_2.py).

By this way, we shall avoid having the conflict because every engineer is working on a unique file and hence it wont create any conflict.

**Comment by Parumugam** [comment]
@mattsm @SyHuangCBN , Can we please think of a solution for this?

because, at the moment, we started getting many Merge Conflict as multiple people work on same file. When people commit their code, its a clean master file, but during the intreval of Code Review, if some others code gets reviewed and merged into the master file, the other people changes becomes merge conflict and they have to rebase their code. 

But the same problem continues even after rebase. Someone else code gets merged again and the rebase has to be done again.

So we might lost a significant amount of time in rebasing the code. 

If we can have one file per testcase and if we can maintain those files inside folders, it will avoid the Merge conflicts, but I understand, its more work for boardfarm to identify the testcase. 

So, can we please think of a method to solve this merge conflict problem?

**Comment by mattsm** [comment]
You don't *need* folders you can just use another file as well.

**Comment by Parumugam** [comment]
myself & @SyHuangCBN  discussed about having files like [Component]_[priority].py (Example: management_p1.py), but again, this will not solve the merge conflict as more testcases will be in P2/P3 priority and high chances that people will be working on parallel on same priority testcases.

Thats why thought of having a folder structure, so every engineer can work on their unique file so they wont get a merge conflict.

**Comment by Parumugam** [comment]
If you are referring to one file per testcase, I am thinking about folders, because, its not good to have all the [testcase].py file to created in the default folder which will be messy to manage.

**Comment by Parumugam** [comment]
**Solution Discussed:**

1)	Instead of using the [component].py file (Example: management.py), Every Engineer need to create a file in their name [Engineer_name].py (Example: Pranesh.py) and start coding the testcase in the name of the TESTCASE (Example: NTL7465LG-5650)
2)	The Engineers can commit the code in their file name [Engineer_name].py
3)	Reviewers can review the code and if all good, they can provide +1
4)	Once gerrit gets +2, it will merge the code in the file name [Engineer_name].py to the cbn repository
5)	Every morning, The Administrator in Taiwan will MOVE the changes in the each [Engineer_name].py file to respective [Component].py file before any Automation Engineer starts their Automation work for the day.
6)	So Every Morning, [Engineer_name].py file will be empty in cbn repository. So every following commit will not have any “Merge Conflict”.
7)	Incase [Engineer_name].py is not empty in cbn repository, the respective engineer need to check with the Administrator before they start their work for the day.
8)	Incase, if any engineer is working on a file which is a common library, they need to broadcast about their work in “boardfarm” channel so that no other engineer shall work on the same file (or wait for the merge), so that we shall avoid “merge conflict” on common libraries as well.

By following this way, as [Engineer_name].py will be used ONLY by the respective engineer, we can totally remove the “Merge Conflict” possibilities.

**Comment by mattsm** [comment]
Let's avoid the every morning terminology. Let's say as soon as possible. Also, I'm removing this issue since it's not something pertinent to public discussion.

**Comment by mattsm** [comment]
Last comment removed, please discuss privately via email.

**Comment by Parumugam** [comment]
This is not required as we have agreed on the single file (<testcase_name>.py) to avoid the Merge Conflict. Hence closing this issue.
